Feature/ct 2290 consume entity relation updates

nozoz opened the pull request:
This PR primarily enables consuming relation updates from the entity topic. It caches incoming relation updates and attaches them to incoming resources. On cache eviction, it triggers a relation request, prompting the relation service to send updates to affected resources.

nozoz committed:
chore: remove unnecessary logging

nozoz committed:
refactor: use consumer header as true state

nozoz committed:
config: disable opa locally

nozoz committed:
test: add shared configs

nozoz committed:
build: add test implementations & unit and integration test commands

nozoz committed:
fix: use links instead of linksIfPresent to handle empty map of links

nozoz committed:
test: add RelationService integration test

nozoz committed:
test: add RelationService unit test

nozoz committed:
test: add testing config & kafka utility

nozoz committed:
test: move packages

nozoz committed:
chore: remove unused constant

nozoz committed:
CI: call check to run both integration and unit tests

nozoz committed:
feat: add RelationPoolService to queue requests with no present resource

nozoz committed:
build: upgrade to springboot 3.5.5

nozoz committed:
refactor: use ApplicationEventPublisher to publish unfinished relation updates

nozoz committed:
test: event listener enqueues update

nozoz committed:
refactor: use MockitoBean instead of deprecated bean

nozoz committed:
fix: dont publish entity unless previousEntity exists

nozoz committed:
test: RelationPoolService integration tests

nozoz committed:
refactor: rename trueState to persisted

nozoz committed:
fix: ensure and purge topic before each test

nozoz committed:
build: add spring kafka test

nozoz committed:
test: setup embedded kafka configuration

nozoz committed:
test: use embedded kafka

nozoz committed:
test: setup embedded kafka config

nozoz committed:
chore: remove TestInstance annotation

nozoz committed:
test: add embedded kafka on most tests

nozoz committed:
test: disable parallel execution

nozoz committed:
ci: remove docker compose installation

nozoz committed:
ci: use correct docker compose cli

nozoz committed:
test: set await to 10 seconds

nozoz committed:
build: set maxParallelForks to 1 and no parallel execution in junit

nozoz committed:
test: disable test due to flakiness, further testing done locally

nozoz committed:
build: upgrade gradle wrapper to 8.14

nozoz committed:
refactor: use allTopicNames instead of deprecated all on kafka Admin client

nozoz committed:
ci: remove 'v' from version

nozoz committed:
refactor: rename entityConsumer to ResourceEntityConsumer

nozoz committed:
refactor: rename KafkaEntity to ResourceKafkaEntity

nozoz committed:
build: upgrade autorelation-lib to 2.2.0

nozoz committed:
feat: add configuration matching logic

nozoz committed:
refactor: utilize consumer config matches configuration method

nozoz committed:
refactor: rename EventCacheConfig to CaffeineCacheConfig

nozoz committed:
chore: add TODO comment about set instead of list usage

nozoz committed:
feat: add relation caffeine cache

nozoz committed:
refactor: remove EntityProducer

nozoz committed:
feat: add RelationCacheService

nozoz committed:
refactor: remove RelationPoolConfig

nozoz committed:
refactor: remove RelationMutationService

nozoz committed:
refactor: remove return types & cache links instead of queueing them

nozoz committed:
fix: use correct field

nozoz committed:
feat: consume relation update entities from Kafka

nozoz committed:
refactor: remove previously removed classes from constructor & use relationService for relation logic

nozoz committed:
refactor: rename class ResourceEntityConsumer

nozoz committed:
test: remove previously removed classes from test

nozoz committed:
refactor: remove previous 'consumer' header check

nozoz committed:
fix: correct syntax

nozoz committed:
feat: add links from LinkBuffer

nozoz committed:
refactor: rename RelationLinkService to LinkBufferService

nozoz committed:
chore: remove unused function

nozoz committed:
build: upgrade to autorelation-lib 3.0.0

nozoz committed:
test: add relationCache mockk

nozoz committed:
refactor: tearDown is now functional

nozoz committed:
refactor: move package

nozoz committed:
refactor: move filtering logic to consumer

nozoz committed:
chore: remove RelationUpdateEventConsumer

nozoz committed:
refactor: rename RelationUpdateEntityConsumer to RelationUpdateConsumer

nozoz committed:
test: create RelationUpdateConsumerTest

nozoz committed:
test: make arguments stricter

nozoz committed:
refactor: remove unused function

nozoz committed:
refactor: rename LinkBufferService to LinkBuffer

nozoz committed:
test: move packages

nozoz committed:
build: upgrade autorelation-lib to 3.1.0

nozoz committed:
refactor: move all relation mutation logic to RelationUpdater

nozoz committed:
test: verify orchestration

nozoz committed:
refactor: convert CompletedFullSyncConsumer to Kotlin & only consume based on correct configuration

nozoz committed:
Rename .java to .kt

nozoz committed:
build: upgrade autorelation-lib to 3.2.1

nozoz committed:
fix: remove entityCreatedTime field

nozoz committed:
fix: remove entityCreatedTime field

nozoz committed:
Rename .java to .kt

nozoz committed:
refactor: convert to Kotlin & add consumer to evicted entities

nozoz committed:
refactor: convert resource name to lowercase

nozoz committed:
fix: set Service annotation on CacheEvictionService

nozoz committed:
fix: set correct core-cache version

nozoz committed:
chore: remove test for previously removed class

nozoz committed:
build: upgrade autorelation-lib to 4.0.0

nozoz committed:
fix: remove unused RelationOperation from CacheEvictionService

nozoz committed:
fix: remove unused RelationOperation from publishDeleteRequest method

nozoz committed:
refactor: update relation handling in RelationService for improved link management

nozoz committed:
fix: update relation handling in RelationUpdater to use existing links regarding autorelation-lib upgrade to 4.0.0

nozoz committed:
fix: remove unused lastDelivered parameter from publishDeleteRequest in ResourceService

nozoz committed:
refactor: streamline relation update processing in RelationService

nozoz committed:
refactor: remove persisted property and related method from ResourceKafkaEntity

nozoz committed:
test: remove unused config

nozoz committed:
test: remove unused class

nozoz committed:
test: enhance RelationServiceTest with new relation update scenarios

nozoz committed:
chore: update docker-compose to use latest images and improve service configuration

nozoz committed:
test: format code in ResourceControllerTest for improved readability

nozoz committed:
refactor: remove unused constants from KafkaConstants

nozoz committed:
refactor: update relation mutation methods for improved clarity and functionality

nozoz committed:
feat: attach previous existing links to new resources

nozoz committed:
test: add unit test for updating resource with existing controlled relations

nozoz committed:
test: update RelationServiceTest to reflect changes in relation handling logic

nozoz committed:
test: convert CacheEvictionServiceTest from Java to Kotlin and enhance test cases

nozoz committed:
Rename .java to .kt

nozoz committed:
feat: enable configuration properties scanning and add eviction config

nozoz committed:
test: refactor CacheEvictionServiceTest and integrate eviction config with fixed clock

nozoz committed:
test: add unit test for publishing a relation request per evicted cache object

nozoz committed:
refactor: rename ResourceKafkaEntity to KafkaEntity and simplify class names

nozoz committed:
ci: remove Kafka setup step from CI workflow

nozoz commented on .github/workflows/CD.yaml:
```diff
@@ -33,8 +38,7 @@ jobs:
           tags: |
-            type=ref,event=branch,enable=false
-            type=ref,event=tag,enable=true,format=short,priority=10
+            type=raw,value=${{ env.VERSION }}
```
Blirt gjort for å fjerne 'v' fra image versjonen

nozoz commented on .github/workflows/CI.yaml:
```diff
@@ -12,14 +12,6 @@ jobs:
         uses: actions/checkout@main
 
-      - name: Install Docker Compose
-        run: |
```
Fjernet fordi vi bruker EmbeddedKafka i testene, slik vi slipper å vedlikeholde staten igjennom tester.

nozoz commented on .github/workflows/CI.yaml:
```diff
@@ -30,4 +22,4 @@ jobs:
 
       - name: Build with Gradle
-        run: ./gradlew build
+        run: ./gradlew check
```
check kjører både unit og integrasjonstester (ref: build.gradle)

nozoz replied:
Kommer også til å kjøre Klint i fremtiden

alstad replied:
Endre navn fra "Build with Gradle" til "Check with Gradle"?

nozoz commented on docker-compose.yaml:
```diff
@@ -21,14 +24,18 @@ services:
-    networks: [broker-kafka]
-    depends_on: [kafka]
+  kafka-ui:
+    image: provectuslabs/kafka-ui:latest
```
bruker kafka-ui istedenfor kafdrop

nozoz commented on src/main/java/no/fintlabs/cache/CacheContainer.java:
```diff
@@ -22,7 +22,7 @@ public CacheContainer(ConsumerConfiguration consumerConfig, CacheManager cacheMa
 
     public Cache<FintResource> getCache(String resource) {
-        return resourceCache.get(resource);
+        return resourceCache.get(resource.toLowerCase());
```
alle keys er lagret som lowercase, så da gjør vi en lowercase sjekk på 'laveste' nivå

alstad committed:
Upgraded to the latest 8.* Gradle Wrapper using "./gradlew wrapper --gradle-version 8.14.3"

nozoz committed:
fix: update loop logic in deleteLinks to address potential multiple link removals

nozoz committed:
test: add unit tests for RelationUpdater to verify link operations and update logic

nozoz committed:
refactor: change RelationUpdateConsumer to @Configuration and enhance bean methods

nozoz committed:
refactor: rename deleteEntry to deleteEntity in ResourceService

nozoz committed:
refactor: remove RelationRequestService and update ResourceService to use RelationRequestProducer directly

nozoz committed:
build: update dependencies and remove local Maven repository

nozoz committed:
refactor: replace RelationRequest with createDeleteRequest in CacheEvictionService

nozoz committed:
ci: rename Gradle task in CI workflow to 'check'

daniejoh commented on src/main/java/no/fintlabs/consumer/links/relation/RelationUpdater.kt:
```diff
@@ -0,0 +1,56 @@
+        existingLink.href.endsWith(
+            idLink.href,
+            ignoreCase = true
+        ) // TODO: Replace List<Link> with Set<Link> or improved FintLinks data structure
```
Is this for future development, or meant for this PR?

alstad committed:
chore: correction in logging of retention times

* Milliseconds is abbreviated as "ms", not "-MS".
* Retention time should be written without hyphen between the two words

alstad committed:
chore: reverted naming of CaffeineCacheConfig.java to EventCacheConfig.java.

It is an implementation detail that the configured caches are Caffeine caches. In the context of the app, this class configures event caches

alstad committed:
chore: if-statements shall always use brackets and "timestamp" shall not be abbreviated to "ts" (readability)

alstad committed:
fix: removed specification of test version of FINT resource model libraries because it is not intuitive and because it is silently ignored (overriden by implementation-version)

nozoz committed:
Feature/ct 2301 sync tracker (#5)

* Rename .java to .kt

* refactor: to Kotlin

* refactor: removed persisted field & renamed createdTim to lastModified & add syncType field

* refactor: add syncCorrId

* test: cover missing lastModified and syncCorrId header

* feat: add syncTotalSize field

* test: make syncTotalSize test

* refactor: remove CompletedFullSyncConsumer class

* feat: add getString method to KafkaHeader for UTF-8 decoding

* feat: add new constants for sync-related fields in KafkaConstants

* refactor: only use cache.put that supports lastModified

* fix: remove redundant default value in RelationCacheService constructor

* refactor: update `updateRetentionTime` method to use `ResourceKafkaEntity` as parameter

* refactor: rename MODIFIED_TIME to LAST_MODIFIED in KafkaConstants and ResourceKafkaEntity

* refactor: simplify `ResourceKafkaEntityTest` by introducing helper methods for header setup and entity creation

* test: add unit tests for field mapping and invalid syncType index handling in `ResourceKafkaEntityTest`

* feat: add utility extension functions for Kafka Headers

* refactor: extract sync-related fields into `EntitySync` and update `ResourceKafkaEntity` implementation

* refactor: replace inline `ResourceKafkaEntity` creation with `createResourceKafkaEntity` method

* test: update `ResourceKafkaEntityTest` to align with `EntitySync` changes and refactor field assertions

* test: refactor `ResourceKafkaEntityTest` to use `SyncType` instead of ordinal integers for improved readability and type safety

* refactor: enhance `EntitySync` data class with javadoc

* feat: add `SyncKey` model for tracking sync sessions and utility function for key creation

* feat: add `SyncState` and `FullSyncState` models

* feat: add `SyncPhase` model and derive sync phase logic for `SyncState`

* feat: add `SyncStatus` model and utility function for sync status creation

* Rename .java to .kt

* refactor: rewrite Caffeine cache configuration in Kotlin and introduce `CaffeineCacheProperties` for customizable cache expiration settings

* refactor: tidy `EntityConsumer` for readability by adjusting parameter formatting and method order

* refactor: adjust formatting in `HeaderExtensions` for clarity and maintain consistency

* feat: implement `SyncCache` for managing sync sessions with cache integration and validation logic

* refactor: remove `CompletedFullSyncConsumer` and related eviction delay logic for simplification and performance improvement

* feat: add `completedFullSync` function to `SyncPhase` for checking completed full sync state

* refactor: rename `createResourceKafkaEntity` to `createKafkaEntity` and adjust `KafkaEntity` properties

* feat: implement `SyncStatusProducer` for publishing sync status events

* test: add unit and integration tests for `SyncCache`, `SyncTrackerService`, and Caffeine cache integrations

* fix: remove wrong import

* test: simplify `CacheEvictionServiceTest` by removing unused imports, redundant test cases, and outdated logic

* refactor: simplify `completedFullSync` by removing redundant `SyncPhase` qualifier

* test: rename `ResourceKafkaEntityTest` to `KafkaEntityTest` to match class naming conventions

* test: update `ResourceControllerTest` with `EntitySync` and `SyncType` integration

* test: update `RelationServiceTest` with `EntitySync` and `SyncType` integration

* test: update `SyncCacheTest` and `SyncTrackerServiceTest` with additional sync scenarios and improve code readability

* refactor: inline single-expression functions in `KafkaHeader` for improved readability

* feat: add `retentionTime` field to `KafkaEntity` with nullable header extraction

* feat: integrate `SyncTrackerService` into `ResourceService` for entity sync tracking

* feat: add retention time handling to `CacheService` and integrate with `ResourceService`

* test: add `retentionTime` field as `null` in integration tests for consistency

* refactor: adjust formatting in `updateRetentionTime` method for consistency

* refactor: add TODO comments for aligning status code handling with core logic

* fix: remove unnecessary `.toLowerCase()` calls in `ResourceController` API endpoints

* Rename .java to .kt

* test: convert `ResourceServiceTest` to Kotlin for consistency with updated test framework

* test: add integration tests for `ResourceService` cache eviction and last delivered handling

* fix: integrate `SyncTrackerService` into `ResourceService`

* test: add unit test for overriding full-sync with `totalSize = 1` returning completed

* test: update `SyncCacheIntegrationTest` to verify cache eviction after full-sync completion

* test: increase sleep duration in `ResourceServiceTest` to ensure cache eviction completion

* test: fix flakiness with shared retention state & sleep durations

nozoz committed:
refactor: rename LinkBuffer to UnresolvedLinkCache across the project

nozoz committed:
refactor: rename UnresolvedLinkCache to UnresolvedRelationCache across the project

nozoz committed:
refactor: rename methods in UnresolvedRelationCache and related classes for clarity

nozoz committed:
refactor: make EntitySync nullable

alstad committed:
refactor: moved integration tests into src/integrationTest/java and unit tests into the default src/test/java directories.

This setup follows best-practices and allow unit testing of internal Kotlin functions (was blocked because detected as different module scope)

nozoz committed:
fix: handle nullable Sync in ResourceService to avoid potential null reference error

nozoz committed:
docs: add class-level and method-level KDoc for KafkaEntity

alstad committed:
refactor: rewrote integration test for updating cache retention time through processing of received entities. The test was flaky and failed randomly

alstad committed:
Merge remote-tracking branch 'origin/feature/CT-2290-consume-entity-relation-updates' into feature/CT-2290-consume-entity-relation-updates

alstad committed:
refactor: replaced HeaderExtensions.kt with ConsumerRecordExtension.kt to remove a unecessary level of abstraction

alstad committed:
chore: code cleanup

alstad committed:
chore: removed KafkaHeader.kt (should have been removed in earlier commit)

alstad committed:
refactor: removed duplicated classes and interfaces in fint.model.resource package.

These are now found in no.fintlabs:fint-model-resource

nozoz committed:
build: upgrade to fint version 3.21.0 (#6)

alstad committed:
refactor: renamed CacheEvictionService#triggerEviction to CacheEvictionService#evictExpired for improved readability

alstad committed:
refactor: re-implemented sync-tracker to simply code for future maintainability

alstad committed:
Merge remote-tracking branch 'origin/feature/CT-2290-consume-entity-relation-updates' into feature/CT-2290-consume-entity-relation-updates

alstad committed:
refactor: converted ResourceServiceTest from integration test to unit test because it was flaky and unnecessary slow

alstad committed:
chore: removed unnecessary source sets configuration for org.jetbrains.kotlin.jvm Gradle plugin

alstad committed:
refactor: removed use of vulnerable library org.apache.commons:commons-text.

Replaced with basic string manipulation

alstad committed:
refactor: removed retentionTimeMap and related code in #updateRetentionTime from CacheService, as it did not serve any other purpose than to log updates to cache retention time.

* This type of logging should be placed in the cache and not in a wrapper class.
* The retentionTimeMap did not track the retention time the cache was initialized with and therefore logged misleading info on first retention time update

alstad committed:
refactor: moved FintResources from package no.fint.model.resource to no.fintlabs.model.resource in order to align with the package hierarchy used in the rest of the project

alstad committed:
Rename .java to .kt

alstad committed:
refactor: modernized ConsumerConfiguration.java by converting it from a Java class annotated with @Data and @Configuration to a Kotlin data class annotated with @ConfigurationProperties

alstad committed:
refactor: replaced slow string formatting with string concatenation
